### ISSUE TITLE
Trigger more resurveys in an area if changes was found 

### ISSUE THREAD

**Issue by RubenKelevra** [streetcomplete/StreetComplete#2273]
**Use case**

If a user completes a resurvey quest and the tags has been changed, the change is just limited to this one way-segment. 

But if, say a cycle way has been modified, than it's likely not just a localized change, but it's likely that there has been more changes in the vicinity. 

**Proposed Solution**

After completing a resurvey of a quest, answering with changes SC could search in a say 150m radius for the same quest, and show the resurvey quests, regardless on the age of the data. 

This is obviously limited to the one user, which discovered the change, but it gives the opportunity to quickly confirm additional changes in an area by him/her. 

Every newly confirmed change will obviously pop up additionally quests in its radius, so if the change is along a longer street, the user would be able to change the street on the full length until there's no changes anymore - and he can confirm the non-changed areas as well. 

To makes it quicker for those resurveys to show up, they should be sorted as the first quest on a street to pop up, avoiding that the user first have to solve other tasks for the street segment.

**Comment by westnordost** [comment]
I thought about this. It certainly makes sense. However, there is not really any way to implement that. The only thing that could theoretically be done, with a huge amount of effort and refactoring how the quest download and filtering works, would be to have this work only locally for yourself on your device.

The core problem here is that (the community favoured that) the check-date tags should only be added if nothing changed and also be removed again if something changed. This means that if the surface is re-asked every 8 years and for example the smoothness is re-asked every 4 years, then the surface will actually never be re-asked if the smoothness is answered more frequently on the same element.
We discussed this already, the current method how to mark certain features of elements as "old" is simply a crutch that is used for the absence of an API functionality. So I am not particularly after making it work better with using that crutch (which would be this issue) and pumping huge amounts of effort into this. I'd rather wait or invest in an extension of the functionality of the API.

**Comment by RubenKelevra** [comment]
So a 'there's a change' answer would just alter the local database and fake the date of the data for the same quests to an old date to show them on the map to the user. 

The idea is that it should only work locally on one device. There's no need to push anything to the OSM API. 

It would just mimik an experienced mapper who would notice the difference between real world and the database and follow the road until he has mapped all changes.

**Comment by matkoniecz** [comment]
Note 

> huge amount of effort and refactoring how the quest download and filtering works

part

**Comment by westnordost** [comment]
Yes, that would be the solution I described, but note @matkoniecz comment. Also, it is not as straightforward as you make it out to be, because the OSM data is actually not available in the local database at that point. On download, all OSM data for a given bbox is available, out of this, the quests are generated, and then all OSM elements which do not have a quest attached are dropped. So later on, those OSM elements are not in the local database. This is one thing that would need to be changed.

So all things considered, I will not implement this.

**Comment by RubenKelevra** [comment]
@westnordost wrote:

> Yes, that would be the solution I described, but note @matkoniecz comment. Also, it is not as straightforward as you make it out to be, because the OSM data is actually not available in the local database at that point. On download, all OSM data for a given bbox is available, out of this, the quests are generated, and then all OSM elements which do not have a quest attached are dropped. So later on, those OSM elements are not in the local database. This is one thing that would need to be changed.
> 

I see.

Then the easiest solution I can think of would be, to generate two databases for resurveys:

First filter all resurveys possible out of the OSM data and put it in one database and do the filtering of the time filter on this database, moving those selected by the filter to the regular database. 

This way you can easily move resurveys from one database to the other one without touching most of the code when a change was detected. 

I understand when this is too work intensive at the moment - but maybe worth considering putting it on the backlog instead of rejecting :)

**Comment by RubenKelevra** [comment]
Hey @westnordost I think v32's new database infrastructure would make this feature request possible to solve...

If so, can you reopen it? :)

**Comment by westnordost** [comment]
Hmm, you are right, the data would be there now. Though I still can't think of a way to cleanly implement this.

**Comment by westnordost** [comment]
So, I put some toughts into how this **could** work technically:

1. In order to persist the information that a certain location needs resurvey, there must be a separate database table with positions+radii which require resurvey

2. answering a quest that was created because the data is old instead of not there yet **and** it was answered in a way that suggests the data was changed must be handled in a special way, i.e. not the usual "modify some tags" but go a different route in the code througout, similar to answering "the POI is not there anymore". How exactly this would work (detecting that a quest was created because the data was old) is something I did not think about yet.

3. In that special implementation, additionally to the usual stuff (tagging the changes), an entry to the mentioned database table must be added

4. `MapDataWithEditsSource` would need to listen to changes on the mentioned database table and also, when queried for map data, it will modify the last edit date of any element that is within the radius as persisted in the mentioned database table to 1.1.1970 or something before passing it on to f.e. the component that creates the quests

If these four points are implemented, then all resurvey quests should reappear when one resurvey quest has been answered with "it changed", regardless of last edit date.
However, there are a few problems:

1. When answering those quests, they won't disappear after answering, because the edit date is hacked to be 1970 for all the elements. I didn't think this through yet, but I haven't thought of a solution without strings attached: Either we need yet another database table in which exceptions to this last edit date hackery is recorded... or use a condition like "if edit date of element is newer than just now when I answered that quest, then don't edit the last edit date to 1970". But that won't work just like that because if several things are changed as  onsequence of the resurvey-trigger, the "just now" date will pushed further and further into the present, meaning that the quests you solved just a few minutes ago will reappear. The cut-off date could be set further into the past (f.e. "start of day"), but this does not solve but only mitigates the problem.

2. **All** quests with a resurvey-component will reappear. Thinking about this, this is really not at all what we want. Oh, so the opening hours of that one shop changed, it's open on Saturday one hour longer? Well then let's just resurvey **everything** in the whole neighbourhood - cycleway infrastructure, recycling containers, **all other** opening hours etc etc. - If this would be implemented even in a very good way, it will still feel like a bug

---

**TLDR**: I am closing this issue because it is both a huge amount of effort plus there are some open questions that can likely only be _mitigated_ (not solved) on the technical side and even if implemented, it seems highly inappropriate to solve the connected use case.

As for many cases, if a surveyor notices that more things changed in for example a modernization effort of the street infrastructure, it is better to leave a note. There are a few alternative approaches which mitigate the described use case which are:
1. StreetComplete uses check_date more often, i.e. it does not delete it if it is already there (already implemented for v32.0)
2. The layers approach could help here, as the surveyor will see an overview of how the data is currently tagged. This way he will be able to find mistakes and correct them more easily.
